hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerPrepare.java

```diff
@@ -0,0 +1,326 @@
+    long prepareRequestLogIndex =
+        omResponse.getPrepareResponse().getTxnID();
+
+    // Prepare response processing is included in the snapshot,
+    // giving index of 1.
+    Assert.assertEquals(1, prepareRequestLogIndex);
+    for (OzoneManager om: cluster.getOzoneManagersList()) {
+      // Leader should be prepared as soon as it returns response.
+      if (om == leader) {
+        checkPrepared(om, prepareRequestLogIndex);
+      } else {
+        waitAndCheckPrepared(om, prepareRequestLogIndex);
+      }
+    }
+  }
+
+  /**
+   * Writes data to the cluster via the leader OM, and then prepares it.
+   * Checks that every OM is prepared successfully.
+   */
+  @Test
+  public void testPrepareWithTransactions() throws Exception {
+    MiniOzoneHAClusterImpl cluster = getCluster();
+    OzoneClient ozClient = OzoneClientFactory.getRpcClient(getConf());
+
+    String volumeName = UUID.randomUUID().toString();
+    String bucketName = UUID.randomUUID().toString();
+    ObjectStore store = ozClient.getObjectStore();
+
+    store.createVolume(volumeName);
+    OzoneVolume volume = store.getVolume(volumeName);
+    volume.createBucket(bucketName);
+
+    Set<String> writtenKeys = new HashSet<>();
+    for (int i = 1; i <= 10; i++) {
+      String keyName = keyPrefix + i;
+      writeTestData(store, volumeName, bucketName, keyName);
+      writtenKeys.add(keyName);
+    }
+
+    // Make sure all OMs have logs from writing data, so we can check that
+    // they are purged after prepare.
+    for (OzoneManager om: cluster.getOzoneManagersList()) {
+      LambdaTestUtils.await(timeoutMillis, 1000,
+          () -> logFilesPresentInRatisPeer(om));
+    }
+
+    OzoneManager leader = cluster.getOMLeader();
+    OMResponse omResponse = submitPrepareRequest(leader.getOmRatisServer());
+    // Get the log index of the prepare request.
+    long prepareRequestLogIndex =
+        omResponse.getPrepareResponse().getTxnID();
+
+    // Make sure all OMs are prepared and all OMs still have their data.
+    for (OzoneManager om: cluster.getOzoneManagersList()) {
+      // Leader should be prepared as soon as it returns response.
+      if (om == leader) {
+        checkPrepared(om, prepareRequestLogIndex);
+      } else {
+        waitAndCheckPrepared(om, prepareRequestLogIndex);
+      }
+
+      List<OmKeyInfo> keys = om.getMetadataManager().listKeys(volumeName,
+          bucketName, null, keyPrefix, 100);
+
+      Assert.assertEquals(writtenKeys.size(), keys.size());
+      for (OmKeyInfo keyInfo: keys) {
+        Assert.assertTrue(writtenKeys.contains(keyInfo.getKeyName()));
+      }
+    }
+  }
+
+  /**
+   * Writes data to the cluster.
+   * Shuts down one OM.
+   * Writes more data to the cluster.
+   * Submits prepare as ratis request.
+   * Checks that two live OMs are prepared.
+   * Revives the third OM
+   * Checks that third OM received all transactions and is prepared.
+   * @throws Exception
+   */
```

Comment:
Is this failing test that needs more changes?

Reply:
I could not get this test to work without client side changes. Once those are implemented by @avijayanhwx, we will use them to implement more robust integration tests with failures. This is left here for reference to come back to.